![Logo ITQ](img/imagen1.png)

**Nombre:** Alexis Arcos
<br>
**Fecha:** 01/04/2026

![Logo Python](img/imagen2.png)

<h1 align="center">Anatomía, Evolución e Impacto de las Redes Neuronales Convolucionales en la Visión Artificial</h1>

# Introducción

En la era contemporánea, la inteligencia artificial ha dejado de ser un concepto abstracto de la ciencia ficción para consolidarse como el motor de la revolución tecnológica actual. Dentro del vasto universo del **aprendizaje automático (*Machine Learning*)** y el **aprendizaje profundo (*Deep Learning*)**, uno de los hitos más extraordinarios es la capacidad de otorgar "visión" a las máquinas. Interpretar el mundo visual de manera similar—y en ocasiones superior—al ojo humano es un desafío monumental que ha sido superado gracias a las **Redes Neuronales Convolucionales (CNN, por sus siglas en inglés)**.

La génesis de estas redes no es puramente matemática, sino que encuentra sus raíces en la biología. En 1959, los neurofisiólogos David Hubel y Torsten Wiesel realizaron un experimento crucial con la corteza visual de los gatos, descubriendo que las neuronas visuales responden a estímulos muy específicos, como líneas y bordes orientados en ciertos ángulos, y que estas respuestas se organizan de forma jerárquica. Décadas después, investigadores como Yann LeCun tomaron esta inspiración biológica de "campos receptivos" locales para diseñar algoritmos artificiales capaces de procesar cuadrículas de píxeles.

El presente ensayo tiene como objetivo explorar en profundidad la arquitectura interna de las redes neuronales convolucionales, desglosando matemáticamente y conceptualmente sus capas fundamentales, analizando sus ventajas frente a los modelos clásicos y demostrando su aplicación práctica en un entorno real.

# Desarrollo

## El Problema de las Redes Neuronales Tradicionales frente a las Imágenes

Para comprender la necesidad de las CNN, es imperativo entender por qué las redes neuronales artificiales clásicas (**Multilayer Perceptrons o MLP**) fracasan al procesar imágenes de alta resolución. En una red tradicional, los datos de entrada deben ser transformados en un vector unidimensional (un proceso conocido como ***flattening***). Si tenemos una imagen a color de 1000x1000 píxeles, esta se compone de 3 canales (Rojo, Verde y Azul - RGB), lo que resulta en 3 millones de valores de entrada.

Si la primera capa oculta de una red tradicional tuviera apenas 1,000 neuronas, el modelo requeriría 3 mil millones de pesos (*weights*) solo para esa primera conexión. Esto no solo hace que el costo computacional sea inmanejable y propenso al sobreajuste (***overfitting***), sino que, al "aplanar" la imagen, se destruye por completo la estructura espacial bidimensional. Un píxel que formaba parte del ojo de un rostro pierde su relación geométrica con el píxel contiguo. Las redes convolucionales nacen para solucionar exactamente esto: mantienen la estructura tridimensional de la imagen (ancho, alto y profundidad de color) y reducen drásticamente la cantidad de parámetros requeridos.

## Anatomía de una CNN: La Capa de Convolución

El núcleo operativo de una CNN es la **capa de convolución**. En lugar de conectar cada neurona a todos los píxeles, la red aplica pequeños filtros o núcleos (***kernels***), que suelen ser matrices de 3x3 o 5x5. Este filtro se desliza a lo largo y ancho de la imagen realizando una operación matemática conocida como "producto punto" entre los valores del filtro y los píxeles de la imagen original.

El resultado de este deslizamiento es un **mapa de características (*feature map*)**. Durante las primeras capas de la red, estos filtros actúan como detectores de características de muy bajo nivel: identifican bordes horizontales, bordes verticales, contrastes de luz y colores primarios. Lo verdaderamente revolucionario del aprendizaje profundo es que nosotros, como programadores, no diseñamos estos filtros manualmente. A través del algoritmo de **retropropagación (*backpropagation*)**, la red aprende por sí sola qué valores deben tener esos filtros para extraer la información más útil de la imagen.

## No Linealidad y Agrupación: Funciones de Activación y Pooling

Después de cada operación de convolución, los datos pasan por una función de activación no lineal, siendo la más común la función **ReLU (*Rectified Linear Unit*)**. Matemáticamente, ReLU simplemente convierte todos los valores negativos del mapa de características en cero [$f(x) = \max(0, x)$]. Esto es vital, ya que permite a la red aprender relaciones complejas y no lineales, introduciendo la capacidad de modelar escenarios del mundo real.

Inmediatamente después, entra en juego la **capa de agrupación o *Pooling***. Su objetivo es reducir las dimensiones espaciales (ancho y alto) de los mapas de características sin perder la información semántica más importante. El método dominante es el **Max Pooling**, que examina pequeñas regiones de la imagen (por ejemplo, cuadrículas de 2x2) y se queda únicamente con el valor máximo. Esta operación disminuye drásticamente el peso computacional para las siguientes capas y, además, otorga a la red una propiedad llamada **"invarianza traslacional"**: permite que la red reconozca un objeto sin importar si está perfectamente centrado, desplazado a la izquierda o ligeramente rotado.

## Clasificación Final: Capas Totalmente Conectadas

Una vez que la imagen ha pasado por múltiples bloques de convolución y *pooling*, la red ha extraído un conjunto de características de muy alto nivel. Por ejemplo, si estamos analizando rostros, las primeras capas detectaron bordes, las intermedias detectaron formas geométricas como óvalos, y las últimas capas han ensamblado esas formas para identificar "ojos", "narices" o "bocas".

En esta etapa final, la matriz resultante tridimensional se "aplana" en un vector unidimensional y se introduce en una o varias capas de neuronas densamente conectadas (**Fully Connected Layers**), similares a las redes tradicionales. Estas capas finales actúan como el clasificador principal, sopesando las características extraídas para emitir un juicio probabilístico final mediante funciones como **Softmax** (para clasificaciones multiclase) o **Sigmoid** (para clasificación binaria).

# Evolución Histórica de las Arquitecturas Convolucionales

El estado actual de la visión artificial no se alcanzó de la noche a la mañana. La arquitectura básica de las CNN ha sufrido mutaciones y mejoras a lo largo de las décadas, impulsadas por el aumento del poder computacional (uso de GPUs) y la disponibilidad de bases de datos masivas.

## Los Pioneros: LeNet-5 y AlexNet

El primer gran hito funcional fue **LeNet-5**, introducido por Yann LeCun en 1998. Esta red, relativamente superficial para los estándares actuales (con solo siete capas), fue diseñada específicamente para reconocer caracteres numéricos escritos a mano en códigos postales, sentando las bases del uso de capas de convolución seguidas de *pooling*.

Sin embargo, la verdadera explosión del *Deep Learning* ocurrió en 2012 con la llegada de **AlexNet**. Diseñada por Alex Krizhevsky, esta red compitió en el desafío ImageNet (una base de datos con millones de imágenes de alta resolución) y aplastó a todos los algoritmos clásicos de visión artificial. AlexNet introdujo el uso intensivo de la función de activación ReLU y la técnica de *Dropout* (apagar neuronas aleatoriamente durante el entrenamiento) para evitar que la red memorizara los datos y obligarla a generalizar mejor.

## Redes Profundas y ResNet

A medida que los investigadores intentaban crear redes cada vez más profundas (de 20, 50 o 100 capas) para extraer patrones más sutiles, se encontraron con un problema matemático grave: el desvanecimiento del gradiente. En redes muy profundas, la señal de aprendizaje se diluía tanto al viajar hacia atrás por las capas que las primeras capas dejaban de aprender. 

En 2015, Microsoft introdujo **ResNet (Residual Networks)**, una arquitectura que revolucionó el campo al implementar "conexiones residuales" o atajos (*skip connections*). Estas conexiones permiten que la información salte ciertas capas, asegurando que el gradiente fluya sin problemas por toda la red. Gracias a ResNet, hoy en día es posible entrenar arquitecturas de cientos de capas, algo impensable hace una década.

# El Paradigma del Aprendizaje por Transferencia (Transfer Learning)

Uno de los mayores obstáculos en el desarrollo de software basado en redes neuronales convolucionales es la inmensa cantidad de datos y poder de procesamiento (decenas de tarjetas gráficas trabajando durante semanas) necesarios para entrenar un modelo desde cero. Para solucionar esto, la industria adoptó el modelo de **Aprendizaje por Transferencia (*Transfer Learning*)**.

Este paradigma se basa en la idea de que los patrones visuales básicos que una red aprende en un dominio pueden ser inmensamente útiles en otro. En lugar de inicializar los pesos de la red con valores aleatorios, los desarrolladores toman modelos preentrenados por grandes empresas (como las arquitecturas ResNet o VGG16, entrenadas con millones de imágenes de uso general) y "congelan" sus primeras capas. Puesto que las primeras capas detectan geometrías universales como bordes, sombras y texturas, el desarrollador solo necesita reentrenar (*fine-tuning*) las últimas capas de clasificación con su propio conjunto de datos más pequeño.

Un ejemplo fascinante de esta versatilidad inter-dominio es cómo los patrones visuales complejos aprendidos por un modelo diseñado originalmente para el ámbito médico—por ejemplo, diferenciar entre tejido prostático sano y muestras cancerosas analizando la densidad, área o compacidad celular—pueden transferirse y readaptarse de forma analógica para detectar anomalías en la patología vegetal, aislando así tejido enfermo de tejido sano. Esto democratiza el acceso a la inteligencia artificial de alto nivel, permitiendo desarrollar soluciones robustas con fracciones del costo y tiempo computacional originales.

# Conclusión

En síntesis, las Redes Neuronales Convolucionales han marcado un antes y un después en la historia de las ciencias de la computación. Su arquitectura iterativa de extracción, reducción y clasificación imita de forma elegante los mecanismos biológicos, permitiendo automatizar el reconocimiento de patrones visuales a escalas previamente inimaginables.

Han superado el gran cuello de botella del Machine Learning clásico: la extracción manual de características. Hoy en día, las CNN son el tejido neuronal subyacente que permite a los vehículos autónomos de Tesla reconocer peatones, a los gigantes tecnológicos organizar miles de millones de fotografías, y a la medicina moderna detectar tumores en radiografías con una precisión que rivaliza con la de los mejores radiólogos humanos. El futuro de la visión artificial seguirá evolucionando hacia arquitecturas más eficientes, pero el paradigma convolucional permanecerá como el cimiento sobre el cual se construyó la vista de las máquinas.

# Ejemplo Práctico: Detección de Patologías Agrícolas

Para contrastar la teoría con la práctica, evaluemos un problema crítico en la industria agrónoma: la detección temprana de enfermedades en cultivos, específicamente patologías en hojas de tomate, un ámbito donde el procesamiento de imágenes juega un papel estelar.

## El Enfoque Tradicional vs. El Paradigma CNN

Tradicionalmente, este problema se abordaba mediante el procesamiento digital de imágenes clásico. Un ingeniero de datos tenía que preprocesar la imagen, aplicar filtros de umbralización para separar la hoja del fondo y, lo más tedioso, calcular matemáticamente descriptores geométricos precisos utilizando librerías como Scikit-learn. Era necesario extraer manualmente el área de la lesión, la solidez de la mancha o la compacidad de los píxeles necróticos para luego alimentar un modelo de clasificación clásico (como un Random Forest o una Máquina de Vectores de Soporte).

Con una red convolucional, la extracción de descriptores geométricos manuales queda obsoleta. La red recibe los píxeles crudos y, a través de sus filtros, aprende por su cuenta qué patrones de textura y color diferencian un tejido celular sano de uno afectado por patógenos.

A continuación, se presenta la implementación práctica de esta arquitectura en Python

In [1]:
# Importamos las librerías necesarias de TensorFlow y Keras
import tensorflow as tf
from tensorflow.keras import layers, models

# Definimos una función para crear nuestra arquitectura CNN
def crear_modelo_agricola():
    """
    Construye una Red Neuronal Convolucional para la clasificación binaria 
    de imágenes de hojas de tomate (Sano vs. Enfermo).
    """
    modelo = models.Sequential()

    # Extracción de características de bajo nivel (bordes, colores base)
    # Usamos 32 filtros de 3x3. Input shape asume imágenes RGB redimensionadas a 150x150.
    modelo.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(150, 150, 3)))
    modelo.add(layers.MaxPooling2D((2, 2))) # Reducimos dimensionalidad a la mitad

    # Extracción de características de nivel medio (texturas de necrosis)
    # Aumentamos los filtros a 64 a medida que la imagen se hace más pequeña
    modelo.add(layers.Conv2D(64, (3, 3), activation='relu'))
    modelo.add(layers.MaxPooling2D((2, 2)))

    # Extracción de características complejas (geometría de las manchas)
    modelo.add(layers.Conv2D(128, (3, 3), activation='relu'))
    modelo.add(layers.MaxPooling2D((2, 2)))

    # TRANSICIÓN: Aplanamiento de los mapas de características espaciales
    # Pasamos de un tensor 3D a un vector 1D
    modelo.add(layers.Flatten())

    # Clasificador (Capas densamente conectadas)
    # Capa oculta con 512 neuronas para procesar las características extraídas
    modelo.add(layers.Dense(512, activation='relu'))
    
    # Capa de salida con activación Sigmoid para obtener una probabilidad entre 0 y 1
    # Ej: 0 = Tejido Sano, 1 = Presencia de Patología
    modelo.add(layers.Dense(1, activation='sigmoid'))

    return modelo

# Instanciamos el modelo
cnn_tomates = crear_modelo_agricola()

# Compilamos el modelo definiendo el optimizador y la función de pérdida
cnn_tomates.compile(optimizer='adam',
                    loss='binary_crossentropy',
                    metrics=['accuracy'])

# Imprimimos un resumen de la arquitectura para visualizar parámetros y capas
print("Arquitectura de la CNN para Detección de Patologías en Tomates:")
cnn_tomates.summary()

Arquitectura de la CNN para Detección de Patologías en Tomates:


C:\Users\RYZEN\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 148, 148, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 74, 74, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 72, 72, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 36, 36, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_2 (Conv2D)               │ (None, 34, 34, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_2 (MaxPooling2D)  │ (None, 17, 17, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 36992)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 512)            │    18,940,416 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │           513 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 19,034,177 (72.61 MB)

 Trainable params: 19,034,177 (72.61 MB)

 Non-trainable params: 0 (0.00 B)

# Referencias Bibliográficas

* **Goodfellow, I., Bengio, Y., & Courville, A. (2016).** *Deep Learning*. MIT Press. Recuperado de http://www.deeplearningbook.org
* **Hubel, D. H., & Wiesel, T. N. (1959).** Receptive fields of single neurones in the cat's striate cortex. *The Journal of Physiology*, 148(3), 574-591.
* **LeCun, Y., Bottou, L., Bengio, Y., & Haffner, P. (1998).** Gradient-based learning applied to document recognition. *Proceedings of the IEEE*, 86(11), 2278-2324.
* **TensorFlow.** (2026). *Documentación Oficial de TensorFlow y Keras*. Recuperado de https://www.tensorflow.org/api_docs

# GitHub

Aquí puedes ver todo mi cuaderno digital por si quieres revisarlo desde dentro:
👉 [**machine-learning-introduccion (GitHub)**](https://github.com/alexis77710/machine-learning-introduccion.git)